File: swift/robot-simulator/Sources/RobotSimulator/RobotSimulator.swift
/* Why, oh why, does swift not have basic
 * trig functions in the standard library?
 */
import Accelerate
extension Double {
    func cos() -> Double {
        var out: [Double] = [0]
        var num: Int32 = 1
        vvcos(&out, [self], &num)
        return out.first!
    }
    func sin() -> Double {
        var out: [Double] = [0]
        var num: Int32 = 1
        vvsin(&out, [self], &num)
        return out.first!
    }
}

// Handling direction and orientation

enum Direction: Int {
    case east  =   0
    case north =  90
    case west  = 180
    case south = 270

    init?(_ degrees: Int) {
        self.init(rawValue: degrees)
    }

    func asRadians() -> Double {
        return Double(self.rawValue) * Double.pi * 2 / 360
    }
}

protocol Orientable {
    var bearing: Direction {get set}
    mutating func orient(_ direction: Direction)
    mutating func turnLeft(_ amount: Int)
    mutating func turnRight()
}

extension Orientable {
    mutating func orient(_ direction: Direction) -> Void {
        self.bearing = direction
    }

    mutating func turnLeft(_ amount: Int = 90) -> Void {
        let newDir: Int = (self.bearing.rawValue + amount) % 360
        self.orient( Direction(newDir)! )
    }

    mutating func turnRight() -> Void {
        self.turnLeft(270)
    }
}

// Handling coordinates and moving

protocol Locatable: Orientable {
    var x: Int {get set}
    var y: Int {get set}
    var coordinates: [Int] {get}
    mutating func at(x: Int, y: Int)
    mutating func place(x: Int, y: Int, direction: Direction)
    mutating func advance()
}

extension Locatable {
    var coordinates: [Int] {
        get {return [self.x, self.y]}
    }

    mutating func at(x: Int, y: Int) -> Void {
        self.x = x
        self.y = y
    }

    mutating func place(x: Int, y: Int, direction: Direction) -> Void {
        self.at(x: x, y: y)
        self.orient(direction)
    }

    mutating func advance() -> Void {
        let dx = Int(self.bearing.asRadians().cos())
        let dy = Int(self.bearing.asRadians().sin())
        self.at( x: self.x + dx, y: self.y + dy )
    }
}

// Handling instructions

enum Command: Character {
    case turnLeft  = "L"
    case turnRight = "R"
    case advance   = "A"

    init?(_ rawValue: Character) {
        self.init(rawValue: rawValue)
    }
}

protocol Scriptable: Locatable {
    func instructions(_ script: String) -> [Command]
    mutating func evaluate(_ script: String)
}

extension Scriptable {
    func instructions(_ script: String) -> [Command] {
        return script.compactMap { Command($0) }
    }

    mutating func evaluate(_ script: String) -> Void {
        instructions(script).forEach {
            switch $0 {
                case .turnLeft:  self.turnLeft()
                case .turnRight: self.turnRight()
                case .advance:   self.advance()
            }
        }
    }
}

// The Robot

class SimulatedRobot: Scriptable {
    var x: Int = 0
    var y: Int = 0
    var bearing: Direction = .north
}

File: euphoria/robot-simulator/robot-simulator.ex
namespace robot

include std/math.e

enum X, Y, DIR

type direction(sequence word)
    switch word do
        case "north", "east", "south", "west" then return 1
        case else return 0
    end switch
end type

type robot(sequence data)
    return
        length(data) = 3 and
        integer(data[X]) and
        integer(data[Y]) and
        direction(data[DIR])
end type

public function move(robot r, sequence commands)
    for i = 1 to length(commands) do
        switch commands[i] do
            case 'R' then r = turn_right(r)
            case 'L' then r = turn_left(r)
            case 'A' then r = advance(r)
        end switch
    end for

    return r
end function

function turn_right(robot r)
    switch r[DIR] do
        case "north" then return {r[X], r[Y], "east"}
        case "east"  then return {r[X], r[Y], "south"}
        case "south" then return {r[X], r[Y], "west"}
        case "west"  then return {r[X], r[Y], "north"}
    end switch
end function

function turn_left(robot r)
    switch r[DIR] do
        case "north" then return {r[X], r[Y], "west"}
        case "east"  then return {r[X], r[Y], "north"}
        case "south" then return {r[X], r[Y], "east"}
        case "west"  then return {r[X], r[Y], "south"}
    end switch
end function

function advance(robot r)
    switch r[DIR] do
        case "north" then return {r[X], r[Y] + 1, r[DIR]}
        case "south" then return {r[X], r[Y] - 1, r[DIR]}
        case "east"  then return {r[X] + 1, r[Y], r[DIR]}
        case "west"  then return {r[X] - 1, r[Y], r[DIR]}
    end switch
end function

